drivers/pwm/pwm_numaker.c

```diff
@@ -0,0 +1,574 @@
+
+	EPWM_ClearPeriodIntFlag(epwm, channel);
+
+	capif_base = (nextEdge == EPWM_CAPTURE_INT_FALLING_LATCH) ? EPWM_CAPIF_CFLIF0_Pos
+								  : EPWM_CAPIF_CRLIF0_Pos;
+	next_if_mask = (0x1ul << (capif_base + channel));
+	time_out_cnt = DT_PROP(DT_NODELABEL(sysclk), clock_frequency) / 2; /* 500 ms time-out */
+	LOG_DBG("===[%s] Channel=0x%x, R-Cnt=0x%x, F-Cnt0x%x, CNT-0x%x", __func__, channel,
+		EPWM_GET_CAPTURE_RISING_DATA(epwm, channel),
+		EPWM_GET_CAPTURE_FALLING_DATA(epwm, channel), epwm->CNT[channel]);
+	curr_cnt = (curr_edge == EPWM_CAPTURE_INT_FALLING_LATCH)
+			  ? EPWM_GET_CAPTURE_FALLING_DATA(epwm, channel)
+			  : (uint16_t)EPWM_GET_CAPTURE_RISING_DATA(epwm, channel);
+
+	/* Wait for Capture Next Indicator */
+	while ((epwm->CAPIF & next_if_mask) == 0) {
+		if (EPWM_GetPeriodIntFlag(epwm, channel)) {
+			EPWM_ClearPeriodIntFlag(epwm, channel);
+			period_reloads++;
+		}
+		if (--time_out_cnt == 0) {
+			status = -EAGAIN;
+			goto done;
+		}
+	}
+
+	/* Clear Capture Falling and Rising Indicator */
+	EPWM_ClearCaptureIntFlag(epwm, channel,
+				 EPWM_CAPTURE_INT_FALLING_LATCH | EPWM_CAPTURE_INT_RISING_LATCH);
+
+	/* Get Capture Latch Counter Data */
+	next_cnt = (nextEdge == EPWM_CAPTURE_INT_FALLING_LATCH)
+			  ? (uint16_t)EPWM_GET_CAPTURE_FALLING_DATA(epwm, channel)
+			  : (uint16_t)EPWM_GET_CAPTURE_RISING_DATA(epwm, channel);
+
+	*cycles = (period_reloads * NUMAKER_PWM_RELOAD_CNT) + curr_cnt - next_cnt;
+	LOG_DBG("===[%s] cycles=0x%x, period_reloads=0x%x, CAPIF=0x%x, cur-0x%x ,next-0x%x",
+		__func__, *cycles, period_reloads, epwm->CAPIF, curr_cnt, next_cnt);
+
+done:
+	return status;
+}
+
+static void __pwm_numaker_isr(const struct device *dev, uint32_t st_channel, uint32_t end_channel)
+{
+	const struct pwm_numaker_config *cfg = dev->config;
+	struct pwm_numaker_data *data = dev->data;
+	EPWM_T *epwm = cfg->epwm;
+	struct pwm_numaker_capture_data *capture;
+	uint32_t int_status;
+	uint32_t cap_intsts;
+	uint32_t cycles = 0;
+	int status = 0;
+	int i;
+	uint32_t int_mask = (1UL << st_channel | 1UL << end_channel);
+	uint32_t cap_int_mask =
+		(EPWM_CAPIF_CFLIF0_Msk << st_channel | EPWM_CAPIF_CRLIF0_Msk << st_channel |
+		 EPWM_CAPIF_CFLIF0_Msk << end_channel | EPWM_CAPIF_CRLIF0_Msk << end_channel);
+
+	/* Get Output int status */
+	int_status = (epwm->AINTSTS & int_mask);
+	/* Clear Output int status */
+	if (int_status != 0x00) {
+		epwm->AINTSTS = int_status;
+	}
+
+	/* Get CAP int status */
+	cap_intsts = (epwm->CAPIF & cap_int_mask);
+
+	/* PWM counter is timing critical, to avoid print msg from irq_isr
+	 *  until getting capture cycles.
+	 */
+	LOG_DBG("===[%s] Channel=0x%x, CAPIEN=0x%x, CAPIF=0x%x, capIntMask=0x%x", __func__,
+		st_channel, epwm->CAPIEN, epwm->CAPIF, cap_int_mask);
+	if (cap_intsts != 0x00) { /* Capture Interrupt */
+		/* Clear CAP int status */
+		epwm->CAPIF = cap_intsts;
+
+		/* Rising latch or Falling latch */
+		for (i = st_channel; i <= end_channel; i++) {
+			capture = &data->capture[i];
```

Comment:
How about having another function for the per-level handling? This is starting to get nested quite deeply.

Reply:
OK, I split cap function for each pwm channel.